src/datatype.c

```diff
@@ -216,6 +216,44 @@ unsigned jl_special_vector_alignment(size_t nfields, jl_value_t *t)
     return alignment;
 }
 
+static int jl_layout_isbits(jl_value_t *ty)
+{
+    if (jl_isbits(ty) && jl_is_leaf_type(ty)) {
+        if (((jl_datatype_t*)ty)->layout) // layout check handles possible layout recursion
+            return 1;
+    }
+    return 0;
+}
+
+static unsigned union_isbits(jl_value_t *ty, size_t *nbytes, size_t *align)
+{
+    if (jl_is_uniontype(ty)) {
+        unsigned na = union_isbits(((jl_uniontype_t*)ty)->a, nbytes, align);
+        if (na == 0)
+            return 0;
+        unsigned nb = union_isbits(((jl_uniontype_t*)ty)->b, nbytes, align);
+        if (nb == 0)
+            return 0;
+        return na + nb;
+    }
+    if (jl_layout_isbits(ty)) {
+        size_t sz = jl_datatype_size(ty);
+        size_t al = ((jl_datatype_t*)ty)->layout->alignment;
+        if (*nbytes < sz)
+            *nbytes = sz;
+        if (*align < al)
```

Comment:
Right, alignments must be powers of N (usually 2), otherwise N * alignment wouldn't be aligned to the alignment. Hence picking the larger alignment is always sufficient.